Auto-indent specific keywords

InspectorMustache opened the pull request:
This automatically indents certain keywords right after they are typed. These keywords are defined in the list `fish-auto-indent-trigger-keywords` as "end", "else" and "case". I think these are the only keywords for which this feature is useful as they are the only ones that can change indentation of an already opened line.

Now, I'm rather new to elisp and Emacs in general so this might not be the most elegant way to achieve this. If you know a better way of implementing this, I would adapt the code accordingly. :)

wwwjfy commented:
Thanks for the PR! And sorry for the late.

I'm not sure about this feature that how helpful it is. Usually, these ending keywords are followed by line breaks, where the indentation is triggered. Triggering it one more time just before that seems not much benefit to me, at the cost of doing regexp matching for each input event which is kind of expensive, though it's trivial when using powerful computing resources nowadays.

I do see a handful of major modes do this, like Python. It is a fancy feature, just may not be worth IMHO.

At the implementation level, this PR looks good, though these can be improved:
- `(require 'subr-x)` is needed to use `string-join`
- using `some` would trigger warning, use `cl-some` instead

If you can make this feature optional, not by default, I'd love to merge.

InspectorMustache commented:
Thanks for the feedback!

I originally wanted this in fish-mode because I use evil and often switch to normal mode/state after entering these keywords. In that case indentation wasn't triggered.

I agree that doing regexp matching on every input event is pretty expensive. I tried to mitigate this by comparing the last character typed to a list of those that could potentially trigger indenting, which should be less expensive. Only after that matches is string-match invoked. That's probably still quite often, though.

Anyway, I incorporated the changes you suggested and added the variable fish-enable-auto-indent which controls whether this feature is used or not. Thanks again!

wwwjfy commented:
Thank you!
